### ISSUE TITLE
A hyperparameter (lambda) and a demo code for CUB dataset

### ISSUE THREAD

**Issue by hoseong-kim** [Elyorcv/SAE#2]
I tried to reproduce your results (61.4%, 60.9%) in SAE paper for CUB dataset, but it fails.
I've only achieved around 40% of CUB dataset (GoogleNet layer pool5 (1024-dim) feature is used.).
Could you let me know a hyperparameter (=lambda) for CUB dataset to reproduce your results (61.4%, 60.9%)? 
And also, could you upload a demo code for CUB dataset?

**Comment by yuanxuanfighting** [comment]
I tried to reproduce your results (65.2%, 65.2%) in SAE paper for SUN dataset, but it fails.
I've only achieved around 51% of SUN dataset
could you upload a demo code for SUN dataset?

**Comment by pallavijog13** [comment]
Can you upload the MAT file for CUB and ImNet-1 dataset?

**Comment by hoseong-kim** [comment]
@pallavijog13,
I've only achieved around 40% of CUB dataset. 
I think that it might be better to wait for authors' upload of MAT file for CUB and ImNet-1 dataset  (#3) instead of mine.

**Comment by pjog** [comment]
@hoseong-kim 
I have a basic question. I am new to Matlab/Octave. I have created cub.mat file using caffe library of Python 3 with the help of caffe.DB and net.forward() call and whatever results I received I stored them in MAT file using savemat() function of Python.scipy.io. Is this the right way? As I am getting very bad results and getting X_te_pro as NaN values. Not sure why? 
I retrieved features from prototxt file through python by changing the data type to ImageData and putting the filenames.txt file for both TEST and TRAIN phases by putting first 150 classes for TRAIN and 151 to 200 classes for TEST phase. The features are extracted from **pool5/7x7_s1** layer.

Am I missing anything? Are we supposed to perform some preprocessing on features before putting it in MAT file? How did you all created CUB.mat file?

**Comment by hoseong-kim** [comment]
Hi @pjog,
I've also used Caffe to extract pool5/7x7_s1 features of Inception. 
There might be no option for that because pre-trained models for Inception network do not exist for PyTorch or MATLAB, etc. Only for Caffe might have a pre-trained model for Inception v1 network.
If you find a pre-trained model for other tools, then please let me know.
Anyway, you can follow my steps to save deep features as ".mat" file.
First, I extracted pool5/7x7_s1 features as ".txt" file by using Caffe for CUB dataset. 
You can refer [this website](http://caffe.berkeleyvision.org/gathered/examples/feature_extraction.html) about feature extraction using Caffe. 
Then, I loaded the ".txt" file with importdata function of MATLAB, and finally saved the data as ".mat" file.

**Comment by pjog** [comment]
Hi @hoseong-kim ,

Yes. I used the method that you have mentioned earlier and got the same results. Just that I performed following steps, few of which might be different than yours:
1. Generated train.txt and test.txt with filepath and label for CUB dataset.
2. Modified train_val.prototxt for ImageData as follows - 

> name: "GoogleNet"
> 
> layer {
>   name: "data"
>   type: "ImageData"
>   top: "data"
>   top: "label"
>   include {
>     phase: TRAIN
>   }
> 
>   transform_param {
>     mirror: true
>     crop_size: 224
>     mean_value: 104
>     mean_value: 117
>     mean_value: 123
>   }
> 
>   image_data_param {
>     source: "/home/user/caffe/examples/googlenet_features/train.txt"
>     batch_size: 50
>     new_height: 256
>     new_width: 256
>   }
> }

3. Run extract_features.bin file with given parameters which generates **LevelDB** database.
4. Extracted **LevelDB** database in numpy array using **Plyvel** library of python (extracted 8822 training and 2965 test features in separate arrays).
5. Generated other arrays from continuous attribute text file of CUB and **param** structure.
6. Pushed all this in MAT file using following code:

> scipy.io.savemat('/home/user/SAE-master/data_zsl/cub.mat', mdict={'X_tr': train_features, 'X_te': test_features, 'S_tr': train_attributes_X, 'S_te_gt': test_attributes, 'S_te_pro': test_attributes, 'param': param})

Another way that I used is:
1. Using caffe library, generate net object

> caffe.set_mode_cpu()
> 
> model_def = caffe_root + 'examples/googlenet_features/train_val.prototxt'
> model_weights = caffe_root + 'models/bvlc_googlenet/bvlc_googlenet.caffemodel'
> 
> net = caffe.Net(model_def,      # defines the structure of the model
>                 model_weights,  # contains the trained weights
>                 caffe.TRAIN)     # use test mode (e.g., don't perform dropout)

2. Run net.forward in loop and extract all the features as follows: (Duplicate code for TEST images)

> counter = 0
> while net.forward(end='pool5/7x7_s1').get('pool5/7x7_s1') != None:
>     value = net.blobs['pool5/7x7_s1'].data
>     #print(len(value))
>     for num_in_batch in range(len(value)):
>         for num_features in range(len(value[num_in_batch])):
>             #print (counter, ",", num_in_batch, ",", num_features)
>             train_features[counter][num_features] = value[num_in_batch][num_features][0][0]
>         counter = counter + 1
>     print (counter)

3. Remaining steps as in earlier method.

I get data correctly in Octave script and with correct dimensions. Just that it is giving very bad accuracy. I really am clueless what is going wrong. Above code is exactly as per given in caffe/examples, just that looping net.forward and passing **end** parameter is something different that I am doing with some trial and error.

Please check my steps and let me know if I am missing something.

Thanks!

**Comment by pjog** [comment]
I think I got the reason why I am not able to get the good result.
Just went through the paper again and found that author has used HIT@5 (K=5) to calculate prediction accuracy. After replacing HITK value from 1 to 5, I got encoder accuracy as 40.8 and decoder accuracy as 57%.
So, now just the problem of lambda parameter remains.

**Comment by hoseong-kim** [comment]
@pjog,
Unfortunately, HIT@5 is only for ImageNet dataset. 
You should use HIT@1 for CUB dataset.

I followed your first method rather than the latter one, but slightly different. 
1. Generated train.txt and test.txt with filepath and label for CUB dataset.
2. Modified train_val.prototxt for CUB dataset to extract features.
3. Run "extract_features" file of Caffe to extract features from Inception network. (The output file is a text file containing extracted features (N x d, where N is the number of data, and d is the dimension of extracted features).
4. Load the text file in MATLAB and save the loaded data as mat file.

**Comment by pjog** [comment]
Ohh right!

Then I think there is a slight difference between our extracted features - your's being better. But since, I am getting the same accuracy as yours in HIT@5, that means my extracted features are close to your features but not same.

I found following 2 links on caffe-users mailing list in which they have mentioned that features extracted using different platforms give different results.

[C++ vs Python feature extraction](https://groups.google.com/forum/#!searchin/caffe-users/feature$20extraction$20transformation/caffe-users/1qMqE_AEySs/aI9Dv_XFFgAJ)

[Matlab features not equal to C++ features: (only) different preprocessing?](https://groups.google.com/forum/#!searchin/caffe-users/feature$20extraction$20transformation/caffe-users/pBOY1_9dWCw/8Et9CUBy1xkJ)

So, possibly some difference between our pre-preprocessing / transformation parameters that you have set in train_val.prototxt file.

For feature extraction to text file, have you used [this](https://github.com/jostosh/caffe-utils) tool?

**Comment by hoseong-kim** [comment]
Actually, when I set to HIT@5 for CUB dataset, I could achieve larger than 60% accuracy.
Your guesses might be possible options I think.
For feature extraction to text file, I utilized a custom code.
I think the tool you introduced might also work.

**Comment by pjog** [comment]
What I mean to say is your HIT@1 and my HIT@5 have same results. Means my correct predictions are coming atleast at top 5 positions which possibly indicate that, as per word vector characteristics, similar words have similar feature vectors. So, possibly, the issue is with preprocessing.

Also, the implementation of Sylvester function in Matlab and Octave (I am running code on Octave) might have some differences in terms of accuracy, which may result in different W. (I may be wrong. Just a guess.)

**Comment by hoseong-kim** [comment]
Okay, I see. 
I hope we can exactly reproduce the results of the paper or the authors provide us with the mat files of other datasets including CUB.

**Comment by Elyorcv** [comment]
Please see the demo for CUB.
